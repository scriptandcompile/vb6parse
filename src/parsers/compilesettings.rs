use num_enum::TryFromPrimitive;
use serde::Serialize;

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum UnroundedFloatingPoint {
    #[default]
    DoNotAllow = 0,
    Allow = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum PentiumFDivBugCheck {
    CheckPentiumFDivBug = 0,
    #[default]
    NoPentiumFDivBugCheck = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum BoundsCheck {
    #[default]
    CheckBounds = 0,
    NoBoundsCheck = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum OverflowCheck {
    #[default]
    CheckOverflow = 0,
    NoOverflowCheck = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum FloatingPointErrorCheck {
    #[default]
    CheckFloatingPointError = 0,
    NoFloatingPointErrorCheck = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum CodeViewDebugInfo {
    #[default]
    NotCreated = 0,
    Created = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum FavorPentiumPro {
    #[default]
    False = 0,
    True = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum Aliasing {
    #[default]
    AssumeAliasing = 0,
    AssumeNoAliasing = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum OptimizationType {
    #[default]
    FavorFastCode = 0,
    FavorSmallCode = 1,
    NoOptimization = 2,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub enum CompilationType {
    // 0
    NativeCode {
        optimization_type: OptimizationType,
        favor_pentium_pro: FavorPentiumPro,
        code_view_debug_info: CodeViewDebugInfo,
        aliasing: Aliasing,
        bounds_check: BoundsCheck,
        overflow_check: OverflowCheck,
        floating_point_check: FloatingPointErrorCheck,
        pentium_fdiv_bug_check: PentiumFDivBugCheck,
        unrounded_floating_point: UnroundedFloatingPoint,
    },
    // -1, default
    PCode,
}
