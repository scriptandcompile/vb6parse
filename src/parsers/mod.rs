//! This module contains the file parsers for the Visual Basic 6 language.
//!
//! The major components of this module are:
//!
//! `VB6Stream` - The main parsing structure for VB6 files. It holds the
//! bytes for a file and provides basic methoda for iterating over the bytes as
//! well as tracking the line, column, and character position of the current byte.
//!
//! `VB6Project` - a structure for holding the parsed information from a VB6
//! project file.
//!
//! `VB6ClassFile` - a structure for holding the parsed information from a VB6
//! class file.
//!
//! `VB6FormFile` - a structure for holding the parsed information from a VB6
//! form file.
//!
//! `VB6ModuleFile` - a structure for holding the parsed information from a VB6
//! module file.
//!

mod header;
mod vb6stream;

pub mod class;
pub mod compilesettings;
pub mod form;
pub mod module;
pub mod objectreference;
pub mod project;
pub mod properties;
pub mod vb6;

pub use class::{
    FileUsage, MtsStatus, Persistance, VB6ClassFile, VB6ClassHeader, VB6ClassProperties,
    VB6ClassVersion,
};

pub use compilesettings::{
    BoundsCheck, CompilationType, FloatingPointErrorCheck, OverflowCheck, PentiumFDivBugCheck,
    UnroundedFloatingPoint,
};
pub use form::{resource_file_resolver, VB6FormFile};
pub use header::{VB6FileAttributes, VB6FileFormatVersion};
pub use module::VB6ModuleFile;

pub use properties::Properties;

pub use objectreference::VB6ObjectReference;

pub use project::{
    CompileTargetType, VB6Project, VB6ProjectClass, VB6ProjectModule, VB6ProjectReference,
};

pub use vb6::{is_english_code, vb6_parse};

pub use vb6stream::VB6Stream;
