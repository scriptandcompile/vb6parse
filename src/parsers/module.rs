use crate::{
    errors::{VB6Error, VB6ErrorKind},
    language::VB6Token,
    parsers::VB6Stream,
    vb6::{keyword_parse, vb6_parse},
};

use serde::Serialize;
use winnow::{
    ascii::{line_ending, space0, space1},
    token::take_until,
    Parser,
};

/// Represents a VB6 module file.
/// A VB6 module file contains a header and a list of tokens.
///
/// The tokens contain the token stream of the code of the class file.
#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub struct VB6ModuleFile<'a> {
    pub name: &'a [u8], // Attribute VB_Name = "Module1"
    pub tokens: Vec<VB6Token<'a>>,
}

impl<'a> VB6ModuleFile<'a> {
    /// Parses a VB6 module file from a byte slice.
    ///
    /// # Arguments
    ///
    /// * `input` The byte slice to parse.
    ///
    /// # Returns
    ///
    /// A result containing the parsed VB6 module file or an error.
    ///
    /// # Errors
    ///
    /// An error will be returned if the input is not a valid VB6 module file.
    ///
    /// # Panics
    ///
    /// This function will panic if the source code is not a valid module file.
    ///
    /// # Example
    ///
    /// ```rust
    /// use vb6parse::parsers::VB6ModuleFile;
    ///
    /// let input = b"Attribute VB_Name = \"Module1\"
    /// Option Explicit
    ///
    /// Private Sub Class_Initialize()
    /// End Sub
    /// ";
    ///
    /// let result = VB6ModuleFile::parse("module.bas".to_owned(), input);
    ///
    /// assert!(result.is_ok());
    /// ```
    pub fn parse(file_name: String, source_code: &'a [u8]) -> Result<Self, VB6Error> {
        let mut input = VB6Stream::new(file_name, source_code);

        match (
            space0,
            keyword_parse("Attribute"),
            space1,
            keyword_parse("VB_Name"),
            space0,
            "=",
            space0,
        )
            .parse_next(&mut input)
        {
            Ok(_) => {}
            Err(e) => {
                return Err(input.error(e.into_inner().unwrap()));
            }
        }

        let name = match (
            "\"",
            take_until(0.., "\""),
            "\"",
            space0::<VB6Stream, VB6ErrorKind>,
            line_ending,
        )
            .take()
            .parse_next(&mut input)
        {
            Ok(name) => name,
            Err(e) => {
                return Err(input.error(e));
            }
        };

        let tokens = match vb6_parse(&mut input) {
            Ok(tokens) => tokens,
            Err(e) => {
                return Err(input.error(e.into_inner().unwrap()));
            }
        };

        Ok(VB6ModuleFile { name, tokens })
    }
}
