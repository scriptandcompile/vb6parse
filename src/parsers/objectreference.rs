use bstr::BStr;
use serde::Serialize;
use uuid::Uuid;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum VB6ObjectReference<'a> {
    Compiled {
        uuid: Uuid,
        version: &'a BStr,
        unknown1: &'a BStr,
        file_name: &'a BStr,
    },
    Project {
        path: &'a BStr,
    },
}

impl Serialize for VB6ObjectReference<'_> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        use serde::ser::SerializeStruct;

        match self {
            VB6ObjectReference::Compiled {
                uuid,
                version,
                unknown1,
                file_name,
            } => {
                let mut state = serializer.serialize_struct("CompiledReference", 4)?;

                state.serialize_field("uuid", &uuid.to_string())?;
                state.serialize_field("version", version)?;
                state.serialize_field("unknown1", unknown1)?;
                state.serialize_field("file_name", file_name)?;

                state.end()
            }
            VB6ObjectReference::Project { path } => {
                let mut state = serializer.serialize_struct("ProjectReference", 1)?;

                state.serialize_field("path", path)?;

                state.end()
            }
        }
    }
}
