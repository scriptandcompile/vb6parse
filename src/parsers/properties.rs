use std::borrow::Cow;
use std::collections::HashMap;
use std::convert::TryFrom;
use std::iter::Iterator;

use bstr::BStr;
use bstr::ByteSlice;
use num_enum::TryFromPrimitive;

use crate::language::StartUpPosition;
use crate::language::VB6Color;

pub struct Properties<'a> {
    key_value_store: HashMap<&'a BStr, Cow<'a, [u8]>>,
}

pub struct PropertiesIter<'a> {
    iter: std::collections::hash_map::Iter<'a, &'a BStr, Cow<'a, [u8]>>,
}

impl Properties<'_> {
    #[must_use]
    pub fn iter(&self) -> PropertiesIter<'_> {
        PropertiesIter {
            iter: self.key_value_store.iter(),
        }
    }
}

impl<'a> Iterator for PropertiesIter<'a> {
    type Item = (&'a BStr, &'a [u8]);

    fn next(&mut self) -> Option<Self::Item> {
        self.iter.next().map(|(key, value)| (*key, value.as_ref()))
    }
}

impl Clone for Properties<'_> {
    fn clone(&self) -> Self {
        Properties {
            key_value_store: self.key_value_store.clone(),
        }
    }
}

impl Default for Properties<'_> {
    fn default() -> Self {
        Properties::new()
    }
}

impl<'a> Properties<'a> {
    #[must_use]
    pub fn new() -> Self {
        Properties {
            key_value_store: HashMap::new(),
        }
    }

    pub fn insert(&mut self, property_key: &'a BStr, value: &'a [u8]) {
        self.key_value_store
            .insert(property_key, Cow::Borrowed(value));
    }

    pub fn insert_resource(&mut self, property_key: &'a BStr, value: Vec<u8>) {
        self.key_value_store.insert(property_key, Cow::Owned(value));
    }

    #[must_use]
    pub fn len(&self) -> usize {
        self.key_value_store.len()
    }

    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.key_value_store.is_empty()
    }

    #[must_use]
    pub fn contains_key(&self, property_key: &BStr) -> bool {
        self.key_value_store.contains_key(property_key)
    }

    #[must_use]
    pub fn get_keys(&self) -> Vec<&'a BStr> {
        self.key_value_store.keys().copied().collect()
    }

    pub fn remove(&mut self, property_key: &BStr) -> Option<Cow<'a, [u8]>> {
        self.key_value_store.remove(property_key)
    }

    pub fn clear(&mut self) {
        self.key_value_store.clear();
    }

    #[must_use]
    pub fn get(&self, property_key: &BStr) -> Option<&[u8]> {
        if !self.key_value_store.contains_key(property_key) {
            return None;
        }

        Some(&self.key_value_store[property_key])
    }

    #[must_use]
    pub fn get_bool(&self, property_key: &BStr, default: bool) -> bool {
        if !self.key_value_store.contains_key(property_key) {
            return default;
        }

        match &self.key_value_store[property_key] {
            Cow::Borrowed(b) => match b {
                &b"0" => false,
                &b"1" | &b"-1" => true,
                _ => default,
            },
            Cow::Owned(b) => match b.as_slice() {
                b"0" => false,
                b"1" | b"-1" => true,
                _ => default,
            },
        }
    }

    #[must_use]
    pub fn get_color(&self, property_key: &BStr, default: VB6Color) -> VB6Color {
        if !self.key_value_store.contains_key(property_key) {
            return default;
        }

        let Ok(property_ascii) = self.key_value_store[property_key].to_str() else {
            // If conversion fails, return default color
            return default;
        };

        match VB6Color::from_hex(property_ascii) {
            Ok(color) => color,
            Err(_) => default,
        }
    }

    #[must_use]
    pub fn get_i32(&self, property_key: &BStr, default: i32) -> i32 {
        if !self.key_value_store.contains_key(property_key) {
            return default;
        }

        let Ok(property_ascii) = self.key_value_store[property_key].to_str() else {
            // If conversion fails, return default value
            return default;
        };

        match property_ascii.parse::<i32>() {
            Ok(value) => value,
            Err(_) => default,
        }
    }

    #[must_use]
    pub fn get_property<T>(&self, property_key: &BStr, default: T) -> T
    where
        T: TryFrom<i32> + TryFromPrimitive,
    {
        if !self.key_value_store.contains_key(property_key) {
            return default;
        }

        let Ok(property_ascii) = self.key_value_store[property_key].to_str() else {
            // If conversion fails, return default value
            return default;
        };

        match property_ascii.parse::<i32>() {
            Ok(value) => T::try_from(value).unwrap_or(default),
            Err(_) => default,
        }
    }

    #[must_use]
    pub fn get_startup_position(
        &self,
        property_key: &BStr,
        default: StartUpPosition,
    ) -> StartUpPosition {
        if !self.key_value_store.contains_key(property_key) {
            return default;
        }

        let Ok(property_ascii) = self.key_value_store[property_key].to_str() else {
            // If conversion fails, return default position
            return default;
        };

        match property_ascii.parse::<i32>() {
            Ok(value) => match value {
                0 => {
                    let client_height = self.get_i32(b"ClientHeight".into(), 3000);
                    let client_width = self.get_i32(b"ClientWidth".into(), 3000);
                    let client_top = self.get_i32(b"ClientTop".into(), 200);
                    let client_left = self.get_i32(b"ClientLeft".into(), 100);

                    StartUpPosition::Manual {
                        client_height,
                        client_width,
                        client_top,
                        client_left,
                    }
                }
                1 => StartUpPosition::CenterOwner,
                2 => StartUpPosition::CenterScreen,
                // 3 is the default value for Windows, but we also want
                // to default to WindowsDefault if the value is not found.
                // I've just commented this out since leaving it in will
                // cause Clippy to complain about the unknown value and the
                // default value in the match arm being the same.
                //
                // 3 => StartUpPosition::WindowsDefault,
                _ => StartUpPosition::WindowsDefault,
            },
            Err(_) => StartUpPosition::WindowsDefault,
        }
    }

    #[must_use]
    pub fn get_option<T>(&'a self, property_key: &'a BStr, default: Option<T>) -> Option<T>
    where
        T: TryFrom<&'a str>,
    {
        if !self.key_value_store.contains_key(property_key) {
            return default;
        }

        let property_ascii = &self.key_value_store[property_key];

        let property_ascii = match property_ascii {
            Cow::Borrowed(b) => match b.to_str() {
                Ok(value) => value,
                Err(_) => return default,
            },
            Cow::Owned(value) => match value.to_str() {
                Ok(value) => value,
                Err(_) => return default,
            },
        };

        match T::try_from(property_ascii) {
            Ok(value) => Some(value),
            Err(_) => default,
        }
    }
}
