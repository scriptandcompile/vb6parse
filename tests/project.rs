use vb6parse::parsers::VB6Project;

#[test]
fn artificial_life_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Artificial-life/Artificial Life.vbp");

    let project = match VB6Project::parse("Artificial Life.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Artificial Life.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn blacklight_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Blacklight-effect/Blacklight.vbp");

    let project = match VB6Project::parse("Blacklight.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Blacklight.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn brightness_effect_part_1_project_load() {
    let project_file_bytes =
        include_bytes!("./data/vb6-code/Brightness-effect/Part 1 - Pure VB6/Brightness.vbp");

    let project = match VB6Project::parse("Brightness.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Brightness.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn brightness_effect_part_2_project_load() {
    let project_file_bytes = include_bytes!(
        "./data/vb6-code/Brightness-effect/Part 2 - API - GetPixel and SetPixel/Brightness2.vbp"
    );

    let project = match VB6Project::parse("Brightness2.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Brightness2.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn brightness_effect_part_3_project_load() {
    let project_file_bytes =
        include_bytes!("./data/vb6-code/Brightness-effect/Part 3 - DIBs/Brightness3.vbp");

    let project = match VB6Project::parse("Brightness3.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Brightness3.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn color_shift_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Color-shift-effect/ShiftColor.vbp");

    let project = match VB6Project::parse("ShiftColor.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'ShiftColor.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn colorize_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Colorize-effect/Colorize.vbp");

    let project = match VB6Project::parse("Colorize.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Colorize.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn contrast_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Contrast-effect/Contrast.vbp");

    let project = match VB6Project::parse("Contrast.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Contrast.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn curves_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Curves-effect/Curves.vbp");

    let project = match VB6Project::parse("Curves.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Curves.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn custom_image_filters_project_load() {
    let project_file_bytes =
        include_bytes!("./data/vb6-code/Custom-image-filters/CustomFilters.vbp");

    let project = match VB6Project::parse("CustomFilters.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'CustomFilters.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn diffuse_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Diffuse-effect/Diffuse.vbp");

    let project = match VB6Project::parse("Diffuse.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Diffuse.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn edge_detection_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Edge-detection/EdgeDetection.vbp");

    let project = match VB6Project::parse("EdgeDetection.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'EdgeDetection.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn emboss_engrave_effect_project_load() {
    let project_file_bytes =
        include_bytes!("./data/vb6-code/Emboss-engrave-effect/EmbossEngrave.vbp");

    let project = match VB6Project::parse("EmbossEngrave.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'EmbossEngrave.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn fill_image_region_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Fill-image-region/Fill_Region.vbp");

    let project = match VB6Project::parse("Fill_Region.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Fill_Region.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn fire_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Fire-effect/FlameTest.vbp");

    let project = match VB6Project::parse("FlameTest.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'FlameTest.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn game_physics_basic_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Game-physics-basic/Physics.vbp");

    let project = match VB6Project::parse("Physics.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Physics.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn gradient_2d_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Gradient-2D/Gradient.vbp");

    let project = match VB6Project::parse("Gradient.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Gradient.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn grayscale_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Grayscale-effect/Grayscale.vbp");

    let project = match VB6Project::parse("Grayscale.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Grayscale.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn hidden_markov_model_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Hidden-Markov-model/HMM.vbp");

    let project = match VB6Project::parse("HMM.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'HMM.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn histograms_advanced_project_load() {
    let project_file_bytes =
        include_bytes!("./data/vb6-code/Histograms-advanced/Advanced Histograms.vbp");

    let project = match VB6Project::parse("Advanced Histograms.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Advanced Histograms.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn histogram_basic_project_load() {
    let project_file_bytes =
        include_bytes!("./data/vb6-code/Histograms-basic/Basic Histograms.vbp");

    let project = match VB6Project::parse("Basic Histograms.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Basic Histograms.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn levels_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Levels-effect/Image Levels.vbp");

    let project = match VB6Project::parse("Image Levels.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Image Levels.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn mandelbrot_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Mandelbrot/Mandelbrot.vbp");

    let project = match VB6Project::parse("Mandelbrot.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Mandelbrot.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn map_editor_2d_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Map-editor-2D/Map Editor.vbp");

    let project = match VB6Project::parse("Map Editor.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Map Editor.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn nature_effects_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Nature-effects/NatureFilters.vbp");

    let project = match VB6Project::parse("NatureFilters.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'NatureFilters.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn randomize_effects_project_load() {
    let project_file_bytes =
        include_bytes!("./data/vb6-code/Randomize-effects/RandomizationFX.vbp");

    let project = match VB6Project::parse("RandomizationFX.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'RandomizationFX.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn scanner_twain_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Scanner-TWAIN/VB_Scanner_Support.vbp");

    let project = match VB6Project::parse("VB_Scanner_Support.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'VB_Scanner_Support.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn screen_capture_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Screen-capture/ScreenCapture.vbp");

    let project = match VB6Project::parse("ScreenCapture.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'ScreenCapture.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn sepia_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Sepia-effect/Sepia.vbp");

    let project = match VB6Project::parse("Sepia.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Sepia.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn threshold_effect_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Threshold-effect/Threshold.vbp");

    let project = match VB6Project::parse("Threshold.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Threshold.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}

#[test]
fn transparency_2d_project_load() {
    let project_file_bytes = include_bytes!("./data/vb6-code/Transparency-2D/Transparency.vbp");

    let project = match VB6Project::parse("Transparency.vbp", project_file_bytes) {
        Ok(project) => project,
        Err(e) => {
            panic!("Failed to parse class file 'Transparency.vbp': {e}");
        }
    };

    insta::assert_yaml_snapshot!(project);
}
