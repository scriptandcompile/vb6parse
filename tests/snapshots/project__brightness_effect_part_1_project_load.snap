---
source: tests/project.rs
expression: project
---
project_type: Exe
references:
  - uuid: 00020430-0000-0000-c000-000000000046
    unknown1:
      - 50
      - 46
      - 48
    unknown2:
      - 48
    path:
      - 46
      - 46
      - 92
      - 46
      - 46
      - 92
      - 46
      - 46
      - 92
      - 46
      - 46
      - 92
      - 87
      - 105
      - 110
      - 100
      - 111
      - 119
      - 115
      - 92
      - 83
      - 121
      - 115
      - 87
      - 79
      - 87
      - 54
      - 52
      - 92
      - 115
      - 116
      - 100
      - 111
      - 108
      - 101
      - 50
      - 46
      - 116
      - 108
      - 98
    description:
      - 79
      - 76
      - 69
      - 32
      - 65
      - 117
      - 116
      - 111
      - 109
      - 97
      - 116
      - 105
      - 111
      - 110
objects: []
modules: []
classes: []
related_documents: []
designers: []
forms:
  - - 66
    - 114
    - 105
    - 103
    - 104
    - 116
    - 110
    - 101
    - 115
    - 115
    - 46
    - 102
    - 114
    - 109
user_controls: []
user_documents: []
other_properties: {}
unused_control_info: Remove
upgrade_controls: Upgrade
res_file_32_path: []
icon_form:
  - 102
  - 114
  - 109
  - 66
  - 114
  - 105
  - 103
  - 104
  - 116
  - 110
  - 101
  - 115
  - 115
startup:
  - 102
  - 114
  - 109
  - 66
  - 114
  - 105
  - 103
  - 104
  - 116
  - 110
  - 101
  - 115
  - 115
help_file_path: []
title:
  - 118
  - 98
  - 66
  - 114
  - 105
  - 103
  - 104
  - 116
  - 110
  - 101
  - 115
  - 115
exe_32_file_name:
  - 118
  - 98
  - 66
  - 114
  - 105
  - 103
  - 104
  - 116
  - 110
  - 101
  - 115
  - 115
  - 46
  - 101
  - 120
  - 101
exe_32_compatible: []
dll_base_address: 285212672
path_32: []
command_line_arguments: ~
name:
  - 86
  - 66
  - 95
  - 66
  - 114
  - 105
  - 103
  - 104
  - 116
  - 110
  - 101
  - 115
  - 115
description: []
debug_startup_component: []
help_context_id:
  - 48
compatibility_mode: NoCompatibility
version_32_compatibility: []
version_info:
  major: 1
  minor: 0
  revision: 18
  auto_increment_revision: 1
  company_name: []
  file_description:
    - 83
    - 97
    - 109
    - 112
    - 108
    - 101
    - 32
    - 101
    - 120
    - 101
    - 99
    - 117
    - 116
    - 97
    - 98
    - 108
    - 101
  copyright:
    - 169
    - 50
    - 48
    - 50
    - 48
    - 32
    - 84
    - 97
    - 110
    - 110
    - 101
    - 114
    - 32
    - 72
    - 101
    - 108
    - 108
    - 97
    - 110
    - 100
    - 32
    - 45
    - 32
    - 119
    - 119
    - 119
    - 46
    - 116
    - 97
    - 110
    - 110
    - 101
    - 114
    - 104
    - 101
    - 108
    - 108
    - 97
    - 110
    - 100
    - 46
    - 99
    - 111
    - 109
  trademark: []
  product_name:
    - 118
    - 98
    - 66
    - 114
    - 105
    - 103
    - 104
    - 116
    - 110
    - 101
    - 115
    - 115
    - 46
    - 101
    - 120
    - 101
  comments:
    - 169
    - 50
    - 48
    - 50
    - 48
    - 32
    - 84
    - 97
    - 110
    - 110
    - 101
    - 114
    - 32
    - 72
    - 101
    - 108
    - 108
    - 97
    - 110
    - 100
    - 32
    - 45
    - 32
    - 119
    - 119
    - 119
    - 46
    - 116
    - 97
    - 110
    - 110
    - 101
    - 114
    - 104
    - 101
    - 108
    - 108
    - 97
    - 110
    - 100
    - 46
    - 99
    - 111
    - 109
server_support_files: Local
conditional_compile: []
compilation_type:
  NativeCode:
    optimization_type: FavorFastCode
    favor_pentium_pro: "True"
    code_view_debug_info: NotCreated
    aliasing: AssumeNoAliasing
    bounds_check: NoBoundsCheck
    overflow_check: NoOverflowCheck
    floating_point_check: NoFloatingPointErrorCheck
    pentium_fdiv_bug_check: NoPentiumFDivBugCheck
    unrounded_floating_point: Allow
start_mode: StandAlone
unattended: Interactive
retained: UnloadOnExit
thread_per_object: 0
threading_model: ApartmentThreaded
max_number_of_threads: 1
debug_startup_option: WaitForComponentCreation
use_existing_browser: Use
property_page: []
